text/0000-let-else.md

```diff
@@ -0,0 +1,473 @@
+and also that the resulting `match` code includes syntax boilerplate (e.g. the block) that could theoretically be reduced today but also interferes with rustfmt's rules:
+
+```rust
+let features = match geojson {
+    GeoJson::FeatureCollection(features) => features,
+    _ => {
+        return Err(format_err_status!(
+            422,
+            "GeoJSON was not a Feature Collection",
+        ));
+    }
+};
+```
+
+However, with if-let this could be more succinct & clear:
+
+```rust
+let GeoJson::FeatureCollection(features) = geojson else {
+    return Err(format_err_status!(
+        422,
+        "GeoJSON was not a Feature Collection",
+    ));
+};
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A common pattern in non-trivial code where static guarantees can not be fully met (e.g. I/O, network or otherwise) is to check error cases when possible before proceeding,
+and "return early", by constructing an error `Result` or an empty `Option`, and returning it before the "happy path" code.
+
+This pattern serves no practical purpose to a computer, but it is helpful for humans interacting with the code.
+Returning early helps improve code clarity in two ways:
+- Ensuring the returned result in near the conditional, visually, as the following logic may be lengthy.
+- Reduces rightward shift, as the error return is now in the block, rather than the following logic.
+
+This RFC proposes _(Rust provides)_ an extension to `let` assignment statements to help with this pattern, an `else { }` which can follow a pattern match
+as a `let` assigning statement:
+
+```rust
+let Some(a) = an_option else {
+    // Called if `an_option` is not `Option::Some(T)`.
+    // This block must diverge (stop executing the existing context to the parent block or function).
+    return;
+};
+
+// `a` is now in scope and is the type which the `Option` contained.
+```
+
+This is a counterpart to `if let` expressions, and the pattern matching works identically, except that the value from the pattern match
+is assigned to the surrounding scope rather than the block's scope.
+
+# Reference-level explanations
+[reference-level-explanation]: #reference-level-explanation
+
+let-else is syntactical sugar for `match` where the non-matched case diverges.
+```rust
+let pattern = expr else {
+    /* diverging expr */
+};
+```
+desugars to
+```rust
+let (each, binding) = match expr { 
+    pattern => (each, binding),
+    _ => { 
+        /* diverging expr */
+    }
+};
+```
+
+Any expression may be put into the expression position except an `if {} else {}` as explain below in [drawbacks][].
+While `if {} else {}` is technically feasible this RFC proposes it be disallowed for programmer clarity to avoid an `... else {} else {}` situation.
+Rust already provides us with such a restriction, [`ExpressionWithoutBlock`][expressions].
+
+Any pattern that could be put into if-let's pattern position can be put into let-else's pattern position.
```

Comment:
```suggestion
Any pattern that could be put into if-let's pattern position can be put into let-else's pattern position.
If the pattern is irrefutable, rustc will emit the `irrefutable_let_patterns` warning lint, as it does with an irrefutable pattern in an `if let`.
```

Reply:
I think that would break destructuring, which is irrefutable pattern matching in `let` statements today.

Reply:
@Fishrock123 That lint would only happen if you write a let-else with an irrefutable pattern, such that the else would become unreachable. Dropping the `else` would eliminate the warning.

I'm suggesting a warning lint for the same reason we have that lint for `if let` rather than an error: sometimes it's convenient for a macro to generate such code, and the macro could allow the lint.

Reply:
Note the reference today says:
> A let statement introduces a new set of variables, given by an irrefutable pattern. 

https://doc.rust-lang.org/reference/statements.html#let-statements

Reply:
Also note there is an [allow_if_let_irrefutables RFC](https://github.com/rust-lang/rfcs/blob/master/text/2086-allow-if-let-irrefutables.md).

Reply:
Added this note about `irrefutable_let_patterns`.